### PR TITLE
Generate types and type specs for all generated functions in aws-elixir and aws-erlang

### PR DESCRIPTION
**Pull request by onno-vos-dev**
Triggering the build and for now this should be considered a DRAFT and a Request For Feedback. The code is a mess and needs some cleaning up and combing over but opening it up in order to get feedback from others. 

Generated aws-elixir code as an example: [comparison](https://github.com/aws-beam/aws-elixir/compare/master-snapshot-20240313...add-typespecs-to-all-generated-functions?expand=1)
Generated aws-erlang code as an example: [comparison](https://github.com/aws-beam/aws-erlang/compare/master-snapshot-20240313...add-typespecs-to-all-generated-functions?expand=1)

### CODE REVIEW

File: lib/aws_codegen/util.ex [hunk 1]
```diff
@@ -30,4 +30,182 @@ defmodule AWS.CodeGen.Util do
...
+  def function_argument_type(:erlang, action) do
+    case Map.get(action.input, "target") do
+      "smithy.api#Unit" -> "\#{}"
+      type ->
+        "#{AWS.CodeGen.Name.to_snake_case(String.replace(type, ~r/com\.amazonaws\.[^#]+#/, ""))}()"
+    end
+  end
+
+  def return_type(:elixir, action) do
+    case Map.get(action.output, "target") do
```
Comment (philss) [inline lib/aws_codegen/util.ex:139]:
```suggestion
    case Map.fetch!(action.output, "target") do
```

File: lib/aws_codegen/util.ex [hunk 2]
```diff
@@ -30,4 +30,182 @@ defmodule AWS.CodeGen.Util do
...
+        Enum.join([normal, "{error, any()}" | errors], " |\n    ")
+    end
+  end
+
+  def reserved_type(type) do
+    if type == "node" || type == "term" || type == "function" || type == "reference" do
+      true
+    else
+      false
+    end
```
Comment (philss) [inline lib/aws_codegen/util.ex:208]: What about just using the binary expression here?

```suggestion
    type == "node" || type == "term" || type == "function" || type == "reference"
```

**Review by philss — COMMENTED** [review]
Nice! I gave a first pass and it's looking good.

Please let me know when to review again :)

File: lib/aws_codegen/post_service.ex [hunk 3]
```diff
@@ -137,10 +155,45 @@ defmodule AWS.CodeGen.PostService do
...
+         name: name,
+         type: shape["type"],
+         member: shape["member"],
+         members: shape["members"],
+         min: shape["min"],
+         enum: shape["enum"],
+         is_input: is_input?(shape)
+       }}
+    end)
+    |> Enum.into(%{})
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply lib/aws_codegen/post_service.ex:183]: I keepm forgetting about `Map.new/2` 😅

File: lib/aws_codegen/types.ex [hunk 4]
```diff
@@ -0,0 +1,203 @@
+defmodule AWS.CodeGen.Types do
+
+  # Unfortunately, gotta patch over auto-defining types that already exist in Elixir
+
+  def shape_to_type(:elixir, "String", _) do
+    "String.t()"
+  end
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply lib/aws_codegen/types.ex:7]: I'll make the necessary changes outside of this commit suggestion (as it'd impact most of these functions)

**Comment by aronisstav** [comment]
Solicited opinion: the output comparison of an erlang module looks good enough to me.

I think that it will likely be unreasonably hard to code anything that can "prettify" the "flat", "expanded" generated types, and the point should be more "easier doc generation" than the code itself being small... I didn't check, and I am still not using `aws-erlang` but perhaps one can offer a minified version of the library without types in that case.

Good job! 😄

**Comment by onno-vos-dev** [comment]
> Solicited opinion: the output comparison of an erlang module looks good enough to me.
> 
> I think that it will likely be unreasonably hard to code anything that can "prettify" the "flat", "expanded" generated types, and the point should be more "easier doc generation" than the code itself being small... I didn't check, and I am still not using `aws-erlang` but perhaps one can offer a minified version of the library without types in that case.
> 
> Good job! 😄

I tried to generate the `edocs` (using `rebar3 ex_doc`) for the generated types but apparently `edoc` explodes trying to parse the binaries `<<"BinaryThingy">>` as if it's an XML-tag so I gotta figure out how to deal with that 👍 But then yes, the idea would be to have good docs on the types and hence aid developers in having more information on what's expected as input into the APIs 👍 

Thank you for the comment 🙇‍♂️

File: lib/aws_codegen/post_service.ex [hunk 5]
```diff
@@ -137,10 +155,39 @@ defmodule AWS.CodeGen.PostService do
...
+       }}
+    end)
+  end
+
+  defp is_input?(shape) do
+    if Map.has_key?(shape, "traits") do
+      Map.has_key?(shape["traits"], "smithy.api#input")
+    else
+      true
+    end
```
Comment (LostKobrakai) [inline lib/aws_codegen/post_service.ex:190]:
```suggestion
    !Map.has_key?(shape, "traits") or Map.has_key?(shape["traits"], "smithy.api#input")
```

File: lib/aws_codegen/rest_service.ex [hunk 6]
```diff
@@ -157,6 +161,60 @@ defmodule AWS.CodeGen.RestService do
...
+                  join_parameter_types(action.required_query_parameters, language),
+                  join_parameter_types(action.required_request_header_parameters, language)
+                ]
+            end
+
+          _ ->
+            []
+        end
+    ])
+  end
```
Comment (LostKobrakai) [inline lib/aws_codegen/rest_service.ex:197]:
```suggestion
  def function_parameter_types(action, required_only \\ false)
  
  def function_parameter_types("GET", required_only) do
    language = action.language
    
    get_parameter_types = 
      if required_only do
        [
          join_parameter_types(action.required_query_parameters, language),
          join_parameter_types(action.required_request_header_parameters, language)
        ]
      else
        [
          join_parameter_types(action.query_parameters, language),
          join_parameter_types(action.request_header_parameters, language),
          join_parameter_types(action.request_headers_parameters, language)
        ]
      end
      
    Enum.join([join_parameter_types(action.url_parameters, language) |  get_parameter_types])
  end
  
  def function_parameter_types(action, _required_only) do
    join_parameter_types(action.url_parameters, action.language)
  end
```

**Comment by onno-vos-dev** [comment]
Updated generated code examples that are in the description. 

Main changes pushed are:
* Generate common errors in order to unify types and reduce noise 
  * Thanks Jesper (colleague) for that tip ❤️ 
* Ensure that for Elixir, typedoc examples are considered code blocks 
  * Thanks @LostKobrakai for that tip! ❤️ 

Reception from colleagues, @LostKobrakai and @philss have been good 👍 
I haven't seen anyone yelling "NOOOOO!" on Elixir/Erlang Slack so I'll polish this up over the coming days so we can get this into a mergeable shape 👍 I'm quite happy with how the output looks, just need to polish up some of the cr*ppy code I wrote in this branch 😄 

Screenshot 2024-03-13 at 20 58 20

**Comment by onno-vos-dev** [comment]
- [ ] Get rid of code duplication in `.eex` modules
- [ ] Move some of this parameter types stuff that currently lives in rest_service.ex to types.ex or a similar module

**Comment by onno-vos-dev** [comment]
Declined in favor of #109 due to wrong branch naming.

**Pull request closed without merging**